internal/apps/dop/providers/project_report/cache.go

```diff
@@ -0,0 +1,233 @@
+}
+
+func (c *Cache) Get(key string) interface{} {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if e, ok := c.values[key]; ok {
+		c.increment(e)
+		return e.value
+	}
+	return nil
+}
+
+func (c *Cache) GetOrSet(key string, value func() (interface{}, error)) (interface{}, error) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	e, ok := c.values[key]
+	if ok {
+		c.increment(e)
+		return e.value, nil
+	}
+	// value doesn't exist.
+	v, err := value()
+	if err != nil || v == nil {
+		return nil, err
+	}
+	e = new(cacheEntry)
+	e.key = key
+	e.value = v
+	// The item returned by value() is either newly allocated or was just
+	// read from the DB, therefore we mark it as persisted to avoid redundant
+	// writes or writing empty object. Once the item is invalidated, caller
+	// has to explicitly set it with Set call.
+	e.persisted = true
+	c.values[key] = e
+	c.increment(e)
+	c.len++
+	return v, nil
+}
+
+func (c *Cache) Set(key string, value interface{}) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if e, ok := c.values[key]; ok {
+		// value already exists for key.  overwrite
+		e.value = value
+		e.persisted = true
+		c.increment(e)
+	} else {
+		// value doesn't exist.  insert
+		e = new(cacheEntry)
+		e.key = key
+		e.value = value
+		e.persisted = true
+		c.values[key] = e
+		c.increment(e)
+		c.len++
+	}
+}
+
+func (c *Cache) Delete(key string) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if e, ok := c.values[key]; ok {
+		c.delete(e)
+	}
+}
+
+func (c *Cache) delete(entry *cacheEntry) {
+	delete(c.values, entry.key)
+	c.remEntry(entry.freqNode, entry)
+	c.len--
+}
+
+func (c *Cache) Len() int {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	return c.len
+}
+
+func (c *Cache) Iterate(fn func(k string, v interface{}) error) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	for k, entry := range c.values {
+		if err := fn(k, entry.value); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *Cache) increment(e *cacheEntry) {
+	e.lastAccessTime = time.Now().Unix()
+	currentPlace := e.freqNode
+	var nextFreq int
+	var nextPlace *list.Element
+	if currentPlace == nil {
+		// new entry
+		nextFreq = 1
+		nextPlace = c.freqs.Front()
+	} else {
+		// move up
+		nextFreq = currentPlace.Value.(*listEntry).freq + 1
+		nextPlace = currentPlace.Next()
+	}
+
+	if nextPlace == nil || nextPlace.Value.(*listEntry).freq != nextFreq {
+		// create a new list entry
+		li := new(listEntry)
+		li.freq = nextFreq
+		li.entries = make(map[*cacheEntry]struct{})
+		if currentPlace != nil {
+			nextPlace = c.freqs.InsertAfter(li, currentPlace)
+		} else {
+			nextPlace = c.freqs.PushFront(li)
+		}
+	}
+	e.freqNode = nextPlace
+	nextPlace.Value.(*listEntry).entries[e] = struct{}{}
+	if currentPlace != nil {
+		// remove from current position
+		c.remEntry(currentPlace, e)
+	}
+}
+
+func (c *Cache) remEntry(place *list.Element, entry *cacheEntry) {
+	entries := place.Value.(*listEntry).entries
+	delete(entries, entry)
+	if len(entries) == 0 {
+		c.freqs.Remove(place)
+	}
+}
```

Comment:
How about using a cache library directly? Such as https://github.com/patrickmn/go-cache

Reply:
good idea, i adjusted using the go-cache library directly